src/Monarch/Document.ts

```diff
@@ -7,79 +7,93 @@ import { asap } from 'asap/browser-asap'
 import 'setimmediate'
 
 interface Spec<input, model, message> {
-    init(input: input): model
     input: input
-    update(message: message): (model: model) => model
-    container: HTMLElement
+    init(input: input): model
+    update: (message: message) => (model: model) => model
     view(model: model): VirtualDomTree<message>
+    container: HTMLElement
 }
 
 interface FinishDiffWorkSpec<message> {
     rootVNode: VirtualDomTree<message>
-    rootPathTree: PatchTree
+    rootPatchTree: PatchTree
 }
 
-export function document<input, model, message>({ init, input, update, container, view }: Spec<input, model, message>) {
+function unsafe_document<input, model, message>({ init, input, update, container, view }: Spec<input, model, message>): void {
     const initialModel = init(input)
     const initialVirtualDomTree = view(initialModel)
-    const outputHandlers = OutputHandlersList.nil(dispatchMessage)
+    const outputHandlers = OutputHandlersList.mkNil(dispatchMessage)
     const scheduler = mkScheduler()
     const environment = { scheduler, dispatchDiffWork, finishDiffWork }
 
     let model = initialModel
     let commitedVirtualDomTree = initialVirtualDomTree
     let diffWork: DiffWork<any, any> | undefined = undefined
+    let patchTree: PatchTree | undefined = undefined
 
-    let requestedAsyncRenderId: number | undefined = undefined
-    let requestedAsyncPerformDiffWorkId: number | undefined = undefined
+    let hasRequestedAsyncRendering = false
+    let hasRequestedAsyncPerformDiffWork = false
```

Comment:
This state is redudant. It can be guessed based on the value of the `diffWork` state. When it is not undefined, the async function call has been requested.

Reply:
I need to set `diffWork` before the condition so it's always gonna be not undefined. If I set `diffWork` inside of condition, It's not gonna perform latest diffWork that came through.